### DISCUSSION TITLE
How can I activate a TagHelper when the module is enabled

### DISCUSSION THREAD

**Discussion by hishamco** [OrchardCMS/OrchardCore#15865] [Q&A]
I'm working on a certain module, I created a TagHelper for that, which is registered when the module is enabled, my question is how can I toggle the visibility depending on module activation, something like what I have seen in most modules that are using Liquid filters

**Comment by ns8482e** [comment]
I guess you need `RequireFeatures` on your feature `Startup`

e.g. Following configures Liquid filters from `Contents` module only if feature `OrchardCore.Liquid` is enabled.

https://github.com/OrchardCMS/OrchardCore/blob/815e29753e1455b830bf75886b6f5063abed5cf2/src/OrchardCore.Modules/OrchardCore.Contents/Startup.cs#L210-L220

**Comment by hishamco** [comment]
This will register the liquid filters when the specified feature(s) is enabled, I'm not sure if the TagHelper in the view, and not registered it will render nothing or show an error

Let me try .. thanks

**Comment by sebastienros** [comment]
I am worried with the notion of "visibility" that you mentioned.
What do you expect to happen when a cshtml is run and uses a TagHelper from a feature that is not enabled?

**Comment by hishamco** [comment]
For now I can do the check in the TagHelper, and suppress the output if the module is disabled

**Comment by ns8482e** [comment]
`@addTagHelper` makes tag helper available in view so if you have reference to the `assembly` then the tag helper is available in view/shape. 

So you control the visibility of view/shape not TagHelper.

If a view is for specific `feature`  then only add `TagHelper` related to that feature on that view.

MVC controller (views) can be bound to feature, but shapes are discovered on module.

I would suggest to define base shape in module and add different alternates via `ShapeTableProvider` defined for each feature to extend base shape. Each alternate shape/view for a specific feature will use TagHelpers intend to use for that feature.

**Comment by hishamco** [comment]
> If a view is for specific feature then only add TagHelper related to that feature on that view.

Make sense, but the scenario I have to add it in the layout ;)

**Comment by ns8482e** [comment]
Tb

> scenario I have to add it in the layout ;)

`Layout ` is shape so above mentioned will work for it,

E.g. default will be `Layout`  `Feature1` will show `Layout__FeatureOne` and `Feature2` will show `Layout__FeatureTwo`

Or may be you keep single `Layout`  and inject a `zone` based on enabled feature.

**Comment by hishamco** [comment]
Thanks a lot, I think I need to deep dive into shapes and zones ;)

**Comment by ns8482e** [comment]
Example of this scenario - injecting Zone is `AdminMenu`.

**Comment by hishamco** [comment]
Exactly!!
